slobrok/src/tests/service_map_mirror/service_map_mirror_test.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <vespa/slobrok/server/mock_map_listener.h>
+#include <vespa/slobrok/server/service_map_mirror.h>
+#include <vespa/vespalib/gtest/gtest.h>
+#include <vespa/vespalib/util/stringfmt.h>
+#include <map>
+
+using namespace vespalib;
+using namespace slobrok;
+using vespalib::make_string_short::fmt;
+
+using Map = std::map<vespalib::string, vespalib::string>;
+
+Map dump(const ServiceMapMirror &history) {
+    Map result;
+    for (const auto & entry : history.allMappings()) {
+        result[entry.name] = entry.spec;
+    }
+    return result;
+}
+
+
+void addTo(ServiceMapMirror &target, const ServiceMapping &mapping) {
+    auto cur = target.currentGeneration();
+    std::vector<vespalib::string> removes = {};
+    ServiceMappingList updates = { mapping };
+    auto nxt = cur;
+    nxt.add();
+    MapDiff diff{cur, removes, updates, nxt};
+    target.apply(diff);
+}
+
+void removeFrom(ServiceMapMirror &target, const ServiceMapping &mapping) {
+    auto cur = target.currentGeneration();
+    std::vector<vespalib::string> removes = { mapping.name };
+    ServiceMappingList updates = { };
+    auto nxt = cur;
+    nxt.add();
+    MapDiff diff{cur, removes, updates, nxt};
+    target.apply(diff);
+}
+
```

Comment:
consider also testing diffs that contain multiple things, like remove/update of the same name etc.